### PR TITLE
Refactor process context in preparation of running as a service.

### PR DESCRIPTION
**Pull request by oschaaf**
In preparation of Nighthawk as a service, refactor `ProcessContext` a bit so we can reuse it.
Also adds a new constructor to `OptionsImpl` that will take a proto as an argument.

### CODE REVIEW

File: include/nighthawk/client/process_context.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include "common/event/real_time_system.h"
+
+namespace Nighthawk {
+namespace Client {
+
+/**
+ * Process context is shared between the CLI and grpc service. It is capable of executing
+ * a full Nighthawk test run. Only a single instance is allowed at a time.
+ */
+class ProcessContext {
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: I take it this isn't related to Envoy's recently introduced `ProcessContext`?

File: include/nighthawk/client/process_context.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include "common/api/api_impl.h"
+#include "common/common/logger.h"
+#include "common/event/real_time_system.h"
+
+namespace Nighthawk {
+namespace Client {
+
+/**
+ * Process context is shared between the CLI and grpc service. It is capable of executing
+ * a full Nighthawk test run. Only a single instance is allowed at a time.
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: Why only a single instance at a time? Is this a singleton?

File: include/nighthawk/client/process_context.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+class ProcessContext {
+public:
+  virtual ~ProcessContext() = default;
+
+  /**
+   * @return uint32_t the number of workers that will be used based on configuration
+   * and available hardware.
+   */
+  virtual uint32_t determineConcurrency() const PURE;
+  virtual Envoy::Event::TimeSystem& time_system() PURE;
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: Comments for the other interface methods missing.

File: include/nighthawk/client/process_context.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  virtual Envoy::Api::Impl& api() PURE;
+  virtual Envoy::Stats::Store& store() const PURE;
+
+  /**
+   * @param formatter used to transform output into the desired format.
+   * @return bool true iff execution was successfull.
+   */
+  virtual bool run(OutputFormatter& formatter) PURE;
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: Does the interface need a formatter, or can it return back some object  that is an neutral representation that can then be formatted?

File: include/nighthawk/client/process_context.h (hunk 1 above)
Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: Unrelated indeed, it's a coincidence

File: include/nighthawk/client/process_context.h (hunk 3 above)
Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: Fixed

File: include/nighthawk/client/process_context.h (hunk 4 above)
Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: The formatter was intended to be that neutral ground, but I think it's naming is less then ideal with this refactor. Maybe `OutputCollector` would be better; one can call `.toProto()` on it.

File: source/client/process_context_impl.h [hunk 5]
```diff
@@ -25,6 +25,15 @@
 namespace Nighthawk {
 namespace Client {
 
+constexpr const char* ProcessContextLockFile = "/tmp/nighthawk.lock";
+
+/**
+ * Only a single instance is allowed at a time machine-wide in this implementation.
```
Comment (htuch) [inline source/client/process_context_impl.h]: This one seems a bit strong; can't we have the end user make the call if we give them best practices? I agree we should prevent obviously bad things, but I can imagine situations in which you want to load test without latency measurement and want multiple NHs.

Comment (oschaaf) [reply source/client/process_context_impl.h]: Yeah, so I was in two minds about this myself as well; would adding something like `--instance-name=unique-name` to allow distinct multiple concurrent instances resolve this in a way you would be comfortable with?

Comment (htuch) [reply source/client/process_context_impl.h]: Yep, sounds good. Maybe in a separate PR?

File: include/nighthawk/client/process_context.h (hunk 4 above)
Comment (htuch) [reply include/nighthawk/client/process_context.h]: Yes, `OutputCollector` is better, +1.

File: include/nighthawk/client/process_context.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * @return Envoy::Event::TimeSystem& The time system implementation used by this instance.
+   */
+  virtual Envoy::Event::TimeSystem& time_system() PURE;
+
+  /**
+   * Envoy::Api::Impl& The api implementation used by this instance.
+   */
+  virtual Envoy::Api::Impl& api() PURE;
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: Why do we export these out of the `ProcessContext`? AFAICT, `ProcessContext` main goal is to encapsulate `run` and perform a full run of a workload. Why do we need to have these other features in the interface?

File: include/nighthawk/client/process_context.h (hunk 1 above)
Comment (htuch) [reply include/nighthawk/client/process_context.h]: How come `ProcessContext` and not `Process`? Don't want to bike shed, just trying to understand if this is actually a context or an execution wrapper.

Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: Renamed to `Process` as you are right in that it actually is an execution wrapper

File: include/nighthawk/client/process_context.h (hunk 4 above)
Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: Renamed to `OutputCollector`

File: source/client/process_context_impl.h (hunk 5 above)
Comment (oschaaf) [reply source/client/process_context_impl.h]: SG, created https://github.com/envoyproxy/nighthawk/issues/87 for tracking that new option.

File: include/nighthawk/client/process_context.h (hunk 6 above)
Comment (oschaaf) [reply include/nighthawk/client/process_context.h]: You're right. Cleaned this up, and purged them from the interface.

File: source/client/process_context_impl.cc [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+      dispatcher_(api_.allocateDispatcher()), cleanup_([this] { tls_.shutdownGlobalThreading(); }),
+      benchmark_client_factory_(options), sequencer_factory_(options), options_(options) {
+  machine_lock_ = open(ProcessContextLockFile, O_CREAT | O_RDWR, 0666);
+  if (flock(machine_lock_, LOCK_EX | LOCK_NB)) {
+    throw NighthawkException(
+        "Only a single ProcessContextImpl instance is allowed to exist at a time (machine wide).");
+  }
+  ares_library_init(ARES_LIB_INIT_ALL);
+  Envoy::Event::Libevent::Global::initialize();
+  configureComponentLogLevels(spdlog::level::from_str(options_.verbosity()));
```
[the remark this answers is not in the document]
Comment (oschaaf) [reply source/client/process_context_impl.cc]: Re: `ProcessWide::ProcessWide`:
Nice, sure. Updated to use it.

File: include/nighthawk/client/process.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ */
+class Process {
+public:
+  virtual ~Process() = default;
+
+  /**
+   * @return uint32_t the number of workers that will be used based on configuration
+   * and available hardware.
+   */
+  virtual uint32_t determineConcurrency() const PURE;
```
Comment (htuch) [inline include/nighthawk/client/process.h]: Who uses this outside of `Process`?

File: include/nighthawk/client/process_context.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#pragma once
```
Comment (htuch) [inline include/nighthawk/client/process_context.h]: Is this file stale?

File: source/client/process_context_impl.cc [hunk 10]
```diff
@@ -0,0 +1,238 @@
+#include <sys/file.h>
```
Comment (htuch) [inline source/client/process_context_impl.cc]: `process_impl.cc`?

File: source/client/process_context_impl.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (htuch) [reply source/client/process_context_impl.cc]: Should we remove the flock from this PR?

File: source/client/process_context_impl.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
+#pragma once
```
Comment (htuch) [inline source/client/process_context_impl.h]: `process_impl.h`

File: source/client/process_context_impl.h [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "exe/process_wide.h"
+
+#include "client/benchmark_client_impl.h"
+#include "client/factories_impl.h"
+
+namespace Nighthawk {
+namespace Client {
+
+constexpr const char* ProcessLockFile = "/tmp/nighthawk.lock";
```
Comment (htuch) [inline source/client/process_context_impl.h]: Maybe remove.

File: source/client/process_impl.cc [hunk 13]
```diff
@@ -0,0 +1,239 @@
+#include "client/process_impl.h"
```
Comment (htuch) [inline source/client/process_impl.cc]: Oh, looks like the above file is stale..

**Comment by oschaaf** [comment]
@htuch sorry about the stale `process_context*` files, I removed them.

File: source/client/process_impl.cc [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    throw NighthawkException(
+        "Only a single ProcessImpl instance is allowed to exist at a time (machine wide).");
+  }
+  configureComponentLogLevels(spdlog::level::from_str(options_.verbosity()));
+  tls_.registerThread(*dispatcher_, true);
+}
+
+ProcessImpl::~ProcessImpl() {
+  if (machine_lock_ > 0) {
+    flock(machine_lock_, LOCK_UN);
```
Comment (htuch) [inline source/client/process_impl.cc]: Can we remove and punt to next PR?

File: include/nighthawk/client/process.h (hunk 8 above)
Comment (htuch) [reply include/nighthawk/client/process.h]: Ping on this one :)

Comment (oschaaf) [reply include/nighthawk/client/process.h]: Removed

File: source/client/process_impl.cc (hunk 14 above)
Comment (oschaaf) [reply source/client/process_impl.cc]: Done

**Comment by oschaaf** [comment]
/assign @htuch 
comments addressed

**Pull request merged**